### PR TITLE
boost cmake version for Tensile 4.2

### PR DESCRIPTION
**Pull request by ce1adon**
To issue #960 

2021/6/11 blocker #980

### CODE REVIEW

**Review by junliume — COMMENTED** [review]
@jerryyin  could you comment on the selection of cmake version for libMLIRMIOpen?

File: Dockerfile
Comment (ce1adon) [reply Dockerfile:83]: 3.15.1 is the minimum version for the latest tensile to build. 
3.20.x is not working, too new for latest tensile build.
3.16.x may work but there may be concern about its compatibility with ubuntu 18.04.

Comment (junliume) [reply Dockerfile:83]: Okay. No objections to this PR, just curious how we determine these dependencies and their long term effects on support.

Comment (ce1adon) [reply Dockerfile:83]: Hopefully @pfultz2 can give some comment on this question.

**Review by junliume — APPROVED** [review]
If the rest of the stack are all using cmake 3.15.1 then it is fine with us. Consistency and stability are my only concerns.

**Comment by ce1adon** [comment]
> If the rest of the stack are all using cmake 3.15.1 then it is fine with us. Consistency and stability are my only concerns.

@pfultz2 Hey Paul, I created issue #960 and PR #961 for switching cmake version. Do you have any idea about the impact on MIOpen?

**Comment by pfultz2** [comment]
Ideally, all of our components should use the cmake version in 18.04(with the exception of LLVM), and most of our components follow that. It is concerning that tensile requires such a new version. The entire rocm stack needs to decide the minimum version of cmake we will use and then there should be a process in place which includes all component owners to consider a newer version.

For now, we have no choice but to upgrade the cmake version in order to build tensile, but there should be a larger discussion about what minimum version of cmake we plan to support in rocm.

File: Dockerfile
Comment (jerryyin) [reply Dockerfile:83]: @junliume Sorry I was busy with the TF/MLIR triaging and didn't got a chance to look at this. I'm a bit concerned at this comment: 
```
 3.20.x is not working, too new for latest tensile build.
```

The cmake required to build the latest LLVM usually bump to the latest version. I can forsee this causing conflict between libMLIRMIOpen and Tensile.

**Review by junliume — CHANGES_REQUESTED** [review]
Comments from MLIR, blocking this PR for now

**Comment by jerryyin** [comment]
@junliume I think this PR is good to go with the result [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/180#issuecomment-854843372). @krzysz00 confirmed that 3.15.1 successfully builds/installs libMLIRMIOpen.a so we are good.

On the other hand, long term wise, we will plan on exporting artifact from MLIR CI so MIOpen doesn't have to rebuild every time.

File: Dockerfile
Comment (jerryyin) [reply Dockerfile:83]: There's a possibility that we have to bump it to use 3.15.1 at all scenarios for next MLIR PRs, we'll see.

**Review by jerryyin — APPROVED** [review]

**Review by junliume — DISMISSED** [review]
Unblock it for comments from MLIR.
3.15.1 is still a little bit awkward, while again as long as overall consistency is kept.

P.S. this page is outdated: https://github.com/ROCmSoftwarePlatform/Tensile/wiki/Dependencies

**Comment by krzysz00** [comment]
Y'all could go for any 3.15.x

We're good through 3.20

**Comment by atamazov** [comment]
CI errors.

**Comment by junliume** [comment]
> CI errors.

Strange that boost cmake version can cause Memory access fault

> Memory access fault by GPU node-2 (Agent handle: 0x278a6d0) on address 0x7f6d32400000. Reason: Page not present or supervisor privilege.

**Comment by ce1adon** [comment]
> CI errors.

@atamazov For some reason the CI is no longer print out the configures that failed in conv2d and immed_con2d, which makes the tracking of the issue very difficult. Is it possible to enable printing out of the driver cmd in CI?

**Comment by atamazov** [comment]
@ce1adon 
> @atamazov For some reason the CI is no longer print out the configures that failed in conv2d and immed_con2d, which makes the tracking of the issue very difficult. Is it possible to enable printing out of the driver cmd in CI?

WRT "_Fp16 Hip Tensile All gfx908_" that failed at run `#3`. I think that adding the following:
```
-DMIOPEN_TEST_FLAGS='--verbose --disable-verification-cache'
```
to this line: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/01ff3da68b68b2ffbc31cb681184a8082ac7ce26/Jenkinsfile#L843

should add necessary printing. The line have been added in #556.

---
:red_circle: Please note that we should use `--disable-verification-cache` in all `conv*` tests (I do not remember exact reasoning, but otherwise CI can run into false verification failures).

**Comment by atamazov** [comment]
Please review.

**Comment by ce1adon** [comment]
> @ce1adon @atamazov could you take a look at the CI failure issues?

I've updated the branch to develop with verbose flag. Hopefully it will give some clue.

**Comment by atamazov** [comment]
`test_immed_conv2d` was created some time ago by copy-pasting from `test_conv2d`. At that time there was no valid support for  logging. Later, I added acceptable logging to `test_conv2d`, but `test_immed_conv2d` still lacks it.

Therefore I recommend using MIOPEN_LOG_LEVEL=6 for triaging the issue. Is it reproducible locally?

**Comment by ce1adon** [comment]
Please keep both #7 and #8 build alive. They are testing different stages.

**Comment by ce1adon** [comment]
> @ce1adon I suspect ther ecould be some problem with some other algorithm, Maybe it's worth disabling ALL algorithms except GEMM for MIOpenTensile "Full" stages? This way, the problems in other algorithms won't affect these stages.
> 
> I am looking into ways of implementing this.

Good thoughts. Will try to prove this theory locally.

**Comment by ce1adon** [comment]
As @atamazov  said,  cmd "MIOPEN_DEBUG_HIP_KERNELS=0 MIOPEN_LOG_LEVEL=5 CTEST_PARALLEL_LEVEL=4 MIOPEN_CONV_PRECISE_ROCBLAS_TIMING=0 MIOPEN_DEBUG_CONV_FFT=0 MIOPEN_DEBUG_CONV_DIRECT=0 MIOPEN_DEBUG_CONV_WINOGRAD=0 MIOPEN_DEBUG_CONV_IMPLICIT_GEMM=0 MIOPEN_DEBUG_CONV_SCGEMM=0 make -j check" actually passed the test so the issue is not in GEMM. 

I ran the tests with log level at 6 and got the following error msg. I will share the full log through email.
```
MIOpen(HIP): Info [GetWrwSolutions] 
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 1-26-26-5x5-128-56-56-1-0x0-2x2-1x1-0-NCHW-FP16-W in file /dockerx/MIOpen/src/kernels/gfx90878.HIP.2_12_0.ufdb.txt
MIOpen(HIP): Info2 [FindRecordUnsafe] Key match: 1-26-26-5x5-128-56-56-1-0x0-2x2-1x1-0-NCHW-FP16-W
MIOpen(HIP): Info2 [FindRecordUnsafe] Contents found: miopenConvolutionBwdWeightsAlgoImplicitGEMM:ConvAsmImplicitGemmGTCDynamicWrwXdlops,0.058538,12800,miopenConvolutionBwdWeightsAlgoImplicitGEMM,<unused>;miopenConvolutionBwdWeightsAlgoGEMM:GemmWrwUniversal,0.315882,4326400,miopenConvolutionBwdWeightsAlgoGEMM,<unused>;miopenConvolutionBwdWeightsAlgoDirect:ConvDirectNaiveConvWrw,2.67181,0,miopenConvolutionBwdWeightsAlgoDirect,<unused>
MIOpen(HIP): Info2 [Measure] Db::FindRecord time: 0.485527 ms
MIOpen(HIP): Info2 [GetWrwSolutionWorkspaceSize] solver_id = ConvAsmImplicitGemmGTCDynamicWrwXdlops
MIOpen(HIP): Info [CompileWrwSolution] solver_id = ConvAsmImplicitGemmGTCDynamicWrwXdlops
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 1x26x26x5x5x128x56x56x1xNCHWxFP16x0x0x2x2x1x1x1xW and solver ConvAsmImplicitGemmGTCDynamicWrwXdlops
MIOpen(HIP): Info [ConvolutionWrwImmediate] solver_id = ConvAsmImplicitGemmGTCDynamicWrwXdlops, workspace = 12800
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 1x26x26x5x5x128x56x56x1xNCHWxFP16x0x0x2x2x1x1x1xW and solver ConvAsmImplicitGemmGTCDynamicWrwXdlops
Memory access fault by GPU node-2 (Agent handle: 0x14faf10) on address 0x7f6436e05000. Reason: Page not present or supervisor privilege.
CMake Error at test_test_immed_conv2d.cmake:7 (message):
  Test failed
```

**Comment by atamazov** [comment]
@ce1adon I will provide you with the necessary patch (that disables all but GEMM for conv tests) soon.

**Comment by atamazov** [comment]
@ce1adon Please let me know if you would like me to update this branch directly. I will give you a .diff file otherwise.

**Comment by atamazov** [comment]
@ce1adon Here it is: [test_conv_all_miotensile_only_gemm.diff.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/6639205/test_conv_all_miotensile_only_gemm.diff.txt). Currently it affects test_conv2d/3d and test_immed_conv2d/3d only.

**Comment by ce1adon** [comment]
> @ce1adon Here it is: [test_conv_all_miotensile_only_gemm.diff.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/6639205/test_conv_all_miotensile_only_gemm.diff.txt). Currently it affects test_conv2d/3d and test_immed_conv2d/3d only.

@atamazov This is not going to work since some configures requires workspace that exceeds the memory limitation in GEMM path. Let's pin relevant author to solve the root cause in ConvAsmImplicitGemmGTCDynamicWrwXdlops, unless there are other concerns.

**Comment by atamazov** [comment]
@ce1adon 
> @atamazov This is not going to work since some configures requires workspace that exceeds the memory limitation in GEMM path. 

Please use `MIOPEN_USE_MIOPENTENSILE` from config.h and skip configures that not applicable for GEMM. See how it is done under `#if TEST_DIRECT_SUPPORTED_CONFIG_ONLY` and under `#if WORKAROUND_MI100_ROM37_HIP_COMPILER_CRASH` in conv_common.hpp. I can do that for you, if you wish.

**Comment by ce1adon** [comment]
> @ce1adon
> 
> > @atamazov This is not going to work since some configures requires workspace that exceeds the memory limitation in GEMM path.
> 
> Please use `MIOPEN_USE_MIOPENTENSILE` from config.h and skip configures that not applicable for GEMM. See how it is done under `#if TEST_DIRECT_SUPPORTED_CONFIG_ONLY` and under `#if WORKAROUND_MI100_ROM37_HIP_COMPILER_CRASH` in conv_common.hpp. I can do that for you, if you wish.

Thank you Artem. These are all good thoughts. But I'm afraid I'm not planning to add workarounds to let this test pass. I would prefer to wait for solution to the blocker. Thanks for understanding!

**Comment by atamazov** [comment]
No problem. But please note that "my" solution improves the quality of miopentensile testing. It makes sure that GEMM will be always fully validated. Without it, the "best" solver will be verified.

**Comment by ce1adon** [comment]
> No problem. But please note that "my" solution improves the quality of miopentensile testing. It makes sure that GEMM will be always fully validated. Without it, the "best" solver will be verified.

Sure. This is in fact my original proposal to run miopentensile test one year ago. Since the test rule has been settled, let's follow it.

File: Dockerfile
Comment (atamazov) [reply Dockerfile:81]: Yes. I do not see anything that enforces us to stay with 3.13.4. Let's switch to 3.15 globally.

Comment (atamazov) [reply Dockerfile:81]: #978 (that bumps CMake to 3.15.1 unconditionally) is merged in.

**Comment by atamazov** [comment]
Merge conflicts.

File: Dockerfile
Comment (ce1adon) [reply Dockerfile]: solved

**Review by atamazov — APPROVED** [review]

**Pull request merged**
